cu29_log_runtime/src/lib.rs

```diff
@@ -1,207 +1,248 @@
+use bincode::config::Configuration;
+use bincode::enc::{Encoder, EncoderImpl};
+use bincode::error::EncodeError;
+use bincode::enc::write::Writer;
+use bincode::enc::Encode;
+#[cfg(debug_assertions)]
+use cu29_intern_strs::read_interned_strings;
 
-// The logging system is basically a global queue.
-static QUEUE: OnceCell<Sender<CuLogEntry>> = OnceCell::new();
+static WRITER: OnceLock<(Mutex<Box<dyn WriteStream<CuLogEntry>>>, RobotClock)> = OnceLock::new();
 
-/// The lifetime of this struct is the lifetime of the logger.
-pub struct LoggerRuntime {
-    clock: RobotClock,
-    handle: Option<JoinHandle<()>>,
-    extra_text_logger: Option<ExtraTextLogger>,
-}
+#[cfg(debug_assertions)]
+static EXTRA_TEXT_LOGGER: OnceLock<Option<ExtraTextLogger>> = OnceLock::new();
 
-impl ClockProvider for LoggerRuntime {
-    fn get_clock(&self) -> RobotClock {
-        self.clock.clone()
-    }
-}
+/// The lifetime of this struct is the lifetime of the logger.
+pub struct LoggerRuntime {}
 
 impl LoggerRuntime {
     /// destination is the binary stream in which we will log the structured log.
     /// extra_text_logger is the logger that will log the text logs in real time. This is slow and only for debug builds.
     pub fn init(
-        clock_source: RobotClock,
+        clock: RobotClock,
         destination: impl WriteStream<CuLogEntry> + 'static,
+        #[allow(unused_variables)]
         extra_text_logger: Option<ExtraTextLogger>,
     ) -> Self {
-        if (!cfg!(debug_assertions)) && extra_text_logger.is_some() {
-            eprintln!("Extra text logger is only available in debug builds. Ignoring the extra text logger.");
-        };
-
-        let mut runtime = LoggerRuntime {
-            clock: clock_source,
-            extra_text_logger,
-            handle: None,
-        };
-        let (s, handle) = runtime.initialize_queue(destination);
-        QUEUE
-            .set(s)
-            .expect("Failed to initialize the logger queue.");
-        runtime.handle = Some(handle);
-        runtime
-    }
-
-    fn initialize_queue(
-        &self,
-        mut destination: impl WriteStream<CuLogEntry> + 'static,
-    ) -> (Sender<CuLogEntry>, JoinHandle<()>) {
-        let (sender, receiver) = bounded::<CuLogEntry>(100);
+        let runtime = LoggerRuntime {};
 
-        #[cfg(debug_assertions)]
-        let (index, extra_text_logger) = if let Some(extra) = &self.extra_text_logger {
-            let index = Some(
-                read_interned_strings(extra.path_to_index.as_path())
-                    .expect("Failed to read the interned strings"),
-            );
-            let logger = Some(extra.logger.clone());
-            (index, logger)
+        // If WRITER is already initialized, update the inner value.
+        // This should only be useful for unit testing.
+        if let Some((writer, _)) = WRITER.get() {
+            let mut writer_guard = writer.lock().unwrap();
+            *writer_guard = Box::new(destination);
```

Comment:
We could make a separate LoggerRuntime testing function instead of having this part of the default init function. Something like

```rust
#[cfg(test)]
fn overwrite_writer()
```

Really depends on if you want extra init calls to be ignored or the last one to apply. I figure the former is expected behavior

Reply:
Yeah this one is a little bit dodgy, I believe replace should be the behavior and should basically never happen in prod. I just encountered that in a unit test as we cannot really remove this value, at the same time I don't want to put an Option that will always be tested for a behavior that should never happen lol